testnet/src/node.rs

```diff
@@ -60,52 +75,149 @@ pub struct Node {
 }
 
 fn spawn_peer(mut this: PeerNetwork, p2p_sock: &SocketAddr, rpc_sock: &SocketAddr,
               burn_db_path: String, stacks_chainstate_path: String, 
-              poll_timeout: u64) -> Result<JoinHandle<()>, NetError> {
+              poll_timeout: u64, relay_channel: Sender<RelayerDirective>) -> Result<JoinHandle<()>, NetError> {
     this.bind(p2p_sock, rpc_sock).unwrap();
+    let (mut dns_resolver, mut dns_client) = DNSResolver::new(5);
+    let mut burndb = BurnDB::open(&burn_db_path, true)
+        .map_err(NetError::DBError)?;
+
+    let mut chainstate = StacksChainState::open(
+        false, TESTNET_CHAIN_ID, &stacks_chainstate_path)
+        .map_err(|e| NetError::ChainstateError(e.to_string()))?;
+    
+    let mut mem_pool = MemPoolDB::open(
+        false, TESTNET_CHAIN_ID, &stacks_chainstate_path)
+        .map_err(NetError::DBError)?;
+
     let server_thread = thread::spawn(move || {
         loop {
-            let mut burndb = match BurnDB::open(&burn_db_path, true) {
-                Ok(x) => x,
-                Err(e) => {
-                    warn!("Error while connecting burnchain db in peer loop: {}", e);
-                    thread::sleep(time::Duration::from_secs(1));
-                    continue;
+            let network_result = this.run(&mut burndb, &mut chainstate, &mut mem_pool, None, poll_timeout)
+                .unwrap();
+
+            if let Err(e) = relay_channel.send(RelayerDirective::HandleNetResult(network_result)) {
+                info!("Relayer hang up with p2p channel: {}", e);
+                break;
+            }
+        }
+    });
+
+    let jh = thread::spawn(move || {
+        dns_resolver.thread_main();
+    });
+
+    Ok(server_thread)
+}
+
+fn spawn_miner_relayer(mut relayer: Relayer, local_peer: LocalPeer,
+                       config: Config, keychain_in: &Keychain,
+                       burn_db_path: String, stacks_chainstate_path: String, 
+                       relay_channel: Receiver<RelayerDirective>, dispatcher_channel: Sender<(StacksHeaderInfo, Vec<StacksTransactionReceipt>)>) -> Result<(), NetError> {
+    // Note: the relayer is *the* block processor, it is responsible for writes to the chainstate --
+    //   no other codepaths should be writing once this is spawned.
+    //
+    // the relayer _should not_ be modifying the burndb,
+    //   however, it needs a mut reference to create read TXs.
+    //   should address via #1449
+    let mut burndb = BurnDB::open(&burn_db_path, true)
+        .map_err(NetError::DBError)?;
+
+    let mut chainstate = StacksChainState::open(
+        false, TESTNET_CHAIN_ID, &stacks_chainstate_path)
+        .map_err(|e| NetError::ChainstateError(e.to_string()))?;
+    
+    let mut mem_pool = MemPoolDB::open(
+        false, TESTNET_CHAIN_ID, &stacks_chainstate_path)
+        .map_err(NetError::DBError)?;
+
+    let mut keychain = keychain_in.clone();
```

Comment:
Should the keychain live exclusively in this thread? It's mutating itself when producing VRF(and microblocks keys in a near future), if we have multiple instances, they could go out of sync.
If not, we might have to split it and introduce a new structure.

Reply:
Probably -- I think the thing to do is probably to have two structs -- one handles the "genesis" execution, doing the initial sortition, which then gets destroyed to construct the second structure, which runs the threaded execution.

In this implementation, the first instance of the keychain is not ever touched again after the clone, but it's definitely not good form to keep it around.

Reply:
As of `67a21e304` -- this is implemented -- the initialization node and running node are split up.